File: matlab/pi-exchange/pireceiver.m
% PIRECEIVER script demonstrating MATLAB-HELICS interface
%
% See pisender for usage information
%
% This example attempts to request the next time equal to its
% sim_stop_time, but the corresponding pisender will request intermediate
% times such that the pireceiver will be granted times earlier than
% requested
%
% Note to non-programmers: if unfamiliar, the assert functions simply check
% if the condition is true and if so continues. If false, execution stops.

%% Initialize HELICS library in MATLAB
helicsStartup()

import helics.*;
%% Configuration
deltat = 0.01;  %Base time interval (seconds)
sim_stop_time = 20;

% HELICS options
helics_core_type = 'zmq'; 
fedinitstring = '--federates=1';  % required with current C interface when using separate processes for each federate

%% Provide summary information
helicsversion = helicsGetVersion();

fprintf('PI RECEIVER: Helics version = %s\n', helicsversion)

%% Create Federate Info object that describes the federate properties
fedinfo = helicsCreateFederateInfo();
assert(not(isempty(fedinfo)))

% Set core type from string
helicsFederateInfoSetCoreTypeFromString(fedinfo, helics_core_type);


% Federate init string
helics.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring);


%% Set the message interval (timedelta) for federate. 
% Note:
% HELICS minimum message time interval is 1 ns and by default
% it uses a time delta of 1 second. What is provided to the
% setTimedelta routine is a multiplier for the default timedelta 
% (default unit = seconds).

% Set one message interval
helicsFederateInfoSetTimeProperty(fedinfo,helics_property_time_delta,deltat);
helicsFederateInfoSetIntegerProperty(fedinfo,helics_property_int_log_level,helics_log_level_warning);



%% Actually create value federate
vfed = helicsCreateValueFederate('MATLAB Pi Receiver Federates',fedinfo);
disp('PI RECEIVER: Value federate created');

% Subscribe to PI SENDER's publication (note: published as global)
sub = helicsFederateRegisterSubscription(vfed, 'testA', '');

disp('PI RECEIVER: Subscription registered (testA)');

%% Start execution
try
    helicsFederateEnterExecutingMode(vfed);
    disp('PI RECEIVER: Entering execution mode');
catch e
    error('PI RECEIVER: Failed to enter execution mode (status = x)\n Try running pisender.m first. (or start the broker seperately)');
end

%% Execution Loop
granted_time =- 1;  %Force at least one run

%% Continue execution until end of requested simulation time
while granted_time <= sim_stop_time

    granted_time = helicsFederateRequestTime(vfed, sim_stop_time);

    isupdated = helicsInputIsUpdated(sub);

    if (isupdated == 1)
        value = helicsInputGetDouble(sub);
        fprintf('PI RECEIVER: Received value = %g at time %4.1f from PI SENDER\n', value, granted_time);
    end
end

%% Shutdown
helicsFederateDestroy(vfed);

disp('PI RECEIVER: Federate finalized');

File: c/pi-exchange/pi_receiver.c
/*
Copyright © 2017-2018,
Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC
All rights reserved. See LICENSE file and DISCLAIMER for more details.
*/
static char help[] = " PI RECEIVER: Simple program to demonstrate the usage of HELICS C Interface.\n\
            This example creates a value federate subscribing to the publication \n\
            registered by PI SENDER.\n\n";

#include <ValueFederate.h>
#include <stdio.h>

int main ()
{
    helics_federate_info fedinfo;
    const char *helicsversion;

    const char *fedinitstring = "--federates=1";
    double deltat = 0.01;
    helics_federate vfed;
    helics_input sub;
    helics_time currenttime = 0.0;
	helics_error err= helicsErrorInitialize();
    double value = 0.0;

    helicsversion = helicsGetVersion ();

    printf ("PI RECEIVER: Helics version = %s\n", helicsversion);
    printf ("%s", help);

    /* Create Federate Info object that describes the federate properties */
    fedinfo = helicsCreateFederateInfo ();

    /* Set core type from string */
    helicsFederateInfoSetCoreTypeFromString (fedinfo, "zmq",&err);

    /* Federate init string */
    helicsFederateInfoSetCoreInitString (fedinfo, fedinitstring,&err);

    /* Set the message interval (timedelta) for federate. Note that
       HELICS minimum message time interval is 1 ns and by default
       it uses a time delta of 1 second. What is provided to the
       setTimedelta routine is a multiplier for the default timedelta.
    */
    /* Set one second message interval */
	helicsFederateInfoSetTimeProperty(fedinfo, helics_property_time_period, deltat, &err);

    helicsFederateInfoSetIntegerProperty(fedinfo, helics_property_int_log_level,1,&err);

    /* Create value federate */
    vfed = helicsCreateValueFederate ("TestB Federate",fedinfo,&err);
    printf ("PI RECEIVER: Value federate created\n");
	//free the federateInfo structure
	helicsFederateInfoFree(fedinfo);

    /* Subscribe to PI SENDER's publication */
    sub = helicsFederateRegisterSubscription (vfed, "testA", "",&err);
    printf ("PI RECEIVER: Subscription registered\n");

    /* Enter initialization mode */
	helicsFederateEnterInitializingMode(vfed, &err);
    if (err.error_code == helics_ok)
    {
        printf ("PI RECEIVER: Entered initialization mode\n");
    }
    else
    {
		printf("PI RECEIVER: Failed to Entered initialization mode: %s\n",err.message);
        return (err.error_code);
    }

    /* Enter execution mode */
	helicsFederateEnterExecutingMode(vfed, &err);
    if (err.error_code == helics_ok)
    {
        printf ("PI RECEIVER: Entered execution mode\n");
    }
	else
	{
		printf("PI RECEIVER: Failed to Entered execution mode: %s\n", err.message);
		return (err.error_code);
	}
    while (currenttime < 0.20)
    {
        int isupdated;
		currenttime=helicsFederateRequestTime (vfed, currenttime, &err);

        isupdated = helicsInputIsUpdated (sub);
        if (isupdated!=helics_false)
        {
            /* NOTE: The value sent by sender at time t is received by receiver at time t+deltat */
            value=helicsInputGetDouble (sub, &err);
            printf ("PI RECEIVER: Received value = %4.3f at time %3.2f from PI SENDER\n", value, currenttime);
        }
    }
    helicsFederateDestroy (vfed);
    printf ("PI RECEIVER: Federate finalized\n");
    helicsCloseLibrary ();
    return (0);
}

